engine/src/main/java/org/camunda/bpm/engine/impl/bpmn/deployer/BpmnDeployer.java

```diff
@@ -255,29 +260,59 @@ protected void removeObsoleteTimers(ProcessDefinitionEntity processDefinition) {
     }
   }
 
-  protected void removeObsoleteEventSubscriptions(ProcessDefinitionEntity processDefinition, ProcessDefinitionEntity latestProcessDefinition) {
-    // remove all subscriptions for the previous version
-    if (latestProcessDefinition != null) {
-      EventSubscriptionManager eventSubscriptionManager = getEventSubscriptionManager();
+  protected void removeObsoleteEventSubscriptions(ProcessDefinitionEntity newLatestProcessDefinition, ProcessDefinitionEntity latestProcessDefinition) {
+    if (latestProcessDefinition != null) { // remove all subscriptions for the previous version
+      List<EventSubscriptionEntity> previousSubscriptions = getPreviousSubscriptionEvents(latestProcessDefinition);
+      for (EventSubscriptionEntity eventSubscriptionEntity : previousSubscriptions) {
+        eventSubscriptionEntity.delete();
+      }
+    } else { // check and remove orphan subscriptions if any
+      List<EventSubscriptionEntity> orphanSubscriptions = getOrphanSubscriptionEvents(newLatestProcessDefinition);
+      for (EventSubscriptionEntity eventSubscriptionEntity : orphanSubscriptions) {
+        log.debug("Deleting orphan event subscription for process definition {}: {}", newLatestProcessDefinition.getId(), eventSubscriptionEntity.getEventName());
+        getEventSubscriptionManager().deleteEventSubscription(eventSubscriptionEntity);
+      }
```

Comment:
❌ In case of existing process definitions (p1,p2,p3, delete p2 and p3, end up with orphan), this code won't be executed. So as we discussed, we need to execute it all the time.

Reply:
Indeed, fixed.